pyxem/signals/segments.py

```diff
@@ -0,0 +1,444 @@
+
+        for i in np.arange(num_vectors):
+            gvectors[i] = np.array(vectors[i].copy())
+            vector_indices[i] = np.array([i], dtype=int)
+
+        correlated_segments = np.zeros_like(segments[:1])
+        correlated_vectors = np.array([0.], dtype=object)
+        correlated_vectors[0] = np.array(np.zeros_like(vectors[:1]))
+        correlated_vector_indices = np.array([0], dtype=object)
+        correlated_vector_indices[0] = np.array([0])
+        i = 0
+        pbar = tqdm(total=np.shape(segments)[0])
+        while np.shape(segments)[0] > i:
+            # For each segment, calculate the normalized cross-correlation to
+            # all other segments, and define add_indices for those with a value
+            # above corr_threshold.
+            corr_list = list(map(
+                lambda x: norm_cross_corr(x, template=segments[i]),
+                segments))
+
+            corr_add = list(map(lambda x: x > corr_threshold, corr_list))
+            add_indices = np.where(corr_add)
+            # If there are more add_indices than vector_threshold,
+            # sum segments and add their vectors. Otherwise, discard segment.
+            if (np.shape(add_indices[0])[0] >= vector_threshold and
+                    np.shape(add_indices[0])[0] > 1):
+                new_segment = np.array([np.sum(segments[add_indices],
+                                               axis=0)])
+                if segment_threshold > 1:
+                    segment_check = np.zeros_like(segments[add_indices],
+                                                  dtype=int)
+                    segment_check[np.where(segments[add_indices])] = 1
+                    segment_check = np.sum(segment_check, axis=0, dtype=int)
+                    segment_mask = np.zeros_like(segments[0], dtype=bool)
+                    segment_mask[
+                        np.where(segment_check >= segment_threshold)] = 1
+                    new_segment = new_segment * segment_mask
+                correlated_segments = np.append(correlated_segments,
+                                                new_segment, axis=0)
+                add_indices = add_indices[0]
+                new_vectors = np.array([0], dtype=object)
+                new_vectors[0] = np.concatenate(gvectors[add_indices],
+                                                axis=0).reshape(-1, 2)
+                correlated_vectors = np.append(correlated_vectors,
+                                               new_vectors, axis=0)
+                new_indices = np.array([0], dtype=object)
+                new_indices[0] = np.concatenate(vector_indices[add_indices],
+                                                axis=0).reshape(-1, 1)
+                correlated_vector_indices = np.append(correlated_vector_indices,
+                                                      new_indices, axis=0)
+            elif np.shape(add_indices[0])[0] >= vector_threshold:
```

Comment:
This condition needs a test.

Reply:
Will write a test.

Reply:
pending coverage this was fixed in https://github.com/dnjohnstone/pyxem/pull/5

Reply:
@tinabe the very large number of conditionals here make this hard to read and debug. Can you simplify it? If not, can you write a test that satisfies all the conditions to use this part of the code?

Reply:
I have attempted to add the tests for which I have been tagged here https://github.com/tinabe/pyxem/tree/nanoseg